src/Worker/Core/Shims/TaskOrchestrationContextWrapper.cs

```diff
@@ -195,19 +195,21 @@ public override Task<T> WaitForExternalEvent<T>(string eventName, CancellationTo
 
         // Create a task completion source that will be set when the external event arrives.
         EventTaskCompletionSource<T> eventSource = new();
-        if (this.externalEventSources.TryGetValue(eventName, out IEventSource? existing))
+        if (this.externalEventSources.TryGetValue(eventName, out Queue<IEventSource>? existing))
         {
-            if (existing.EventType != typeof(T))
+            if (existing.Peek().EventType != typeof(T))
```

Comment:
It should never be empty but, let's make this refactor+NPE-safe

```suggestion
            if (existing.Count > 0 && existing.Peek().EventType != typeof(T))
```